SPECS/calamares/calamares.spec

```diff
@@ -217,6 +219,9 @@ install -p -m 644 %{SOURCE53} %{buildroot}%{_sysconfdir}/calamares/mariner-eula
 %{_libdir}/cmake/Calamares/
 
 %changelog
+* Mon Nov 28 2022 Suresh Babu Chalamalasetty <schalam@microsoft.com> - 3.2.11-40
+- Avoid pulling unneeded dependencies by setting AutoReq to no
```

Comment:
RPM is usually pretty good at figuring out correct dependencies. It can get things wrong if we pack test and/or sample files having a dependency on some exotic tool, which is not needed for the core functionality, but then it's better to have RPM ignore only these specific files with `%global __requires_exclude_from` (example in `qt5-qtbase.spec`) instead of adding a blanked `AutoReq` to drop all auto-generated deps.

What were the unneeded packages in this case and what was pulling them in?

Reply:
Mariner 2.0 toolkit is behaving slightly different than 1.0. Even if a dependency is added for BuildRequires it is pulling by default to Requires. To avoid this we need to add AutoReq to no.

Reply:
That is not the case, as far as I could tell. Latest `calamares` package in PMC doesn't require any of the packages from the BRs (I like to run `dnf repoquery --latest-limit 1 --resolve --requires calamares` to get a nice list).

Building with `AutoReq: no` removes the following dependencies (ran the same DNF command but against the built RPM file):
```bash
calamares-libs
glibc
libgcc
libstdc++
qt5-qtbase
qt5-qtbase-gui
```

Especially worrying ones are `calamares-libs` and the `qt5-qtbase*`, since they are not part of Mariner's base images and thus their absence will cause run-time issues for anyone trying to use this package.

I think we're not taking the right approach here and RPM is working as intended identifying the dependencies. We should be using `AutoReq: no` very carefully/never and instead identify the package's components generating these dependencies and going from there. We can move them to a separate subpackage (if they are not required for the core functionality to work properly), try modifying the build options to rely on different packages/disable some features (if acceptable by our users), or look for yet another solution.

Reply:
Verified with local built ISO with these changes and no regressions seen. OS able to install with UI and terminal installer with no issues. Calamares-libs, qt5-qtbase and other runtime dependencies pulled correctly and isolated llvm which is not needed.

Reply:
Ah, yes, I missed the fact that the updated package still brings in `calamares-libs` - I apologize for the confusion.

You're right, installing the updated `calamares` will still install the current dependencies (minus `llvm`) but now indirectly, because `calamares-libs` still has a dependency on them (`AutoReq: no` applies only to the `calamares` package, not the other ones from `calamares.spec`). The binaries from the `calamares` package, however, still need the now-removed dependencies, though. Here's an example for `/usr/bin/calamares` (part of the `calamares` package) and the Qt5 libraries:

![image](https://user-images.githubusercontent.com/24633722/207059756-0795d757-fade-4803-8d48-4517dc08a487.png)

I think that `AutoReq: no` is a too heavy hammer for what we're trying to do here. While `calamares` and the ISO installation works for the scenario you've tested, the current dependency on `llvm` came from somewhere (it's probably not `calamares` itself) and whatever needed it will now be at risk of not functioning properly or at all. And I agree, I'm also curious of @dmcilvaney's opinion.

To summarize: if the goal is to remove the dependency on `llvm`, we should understand what's bringing it in (`dnf repoquery --requires --tree --resolve calamares` should help immensely) and remove that dependency in a surgical manner (move the offending script/binary to a subpackage, remove it completely, change the package's build steps, etc.), while allowing the RPM's automatic dependency resolution to do its work.

Reply:
I've talked offline, agreed that a blanket `AutoReq:no` is too broad. It may seem to work, but I'd be much happier knowing exactly what files are causing this, and fixing those specific files. Using ldd is one way (as shown by Pawel, also see https://www.steventwheeler.com/linux/2018/10/18/debugging-rpm-dependencies.html). We can also set https://rpm-software-management.github.io/rpm/manual/dependency_generators.html `--rpmfcdebug` to debug the generators. I feel like the full llvm package should not be getting pulled, even if some related libraries ARE needed.

Reply:
One more thing I had time to look at: I think the dependency on `llvm` was being brought by `mesa-dri-drivers`. Since that dependency is being dropped from the Qt5 packages in this PR, there's a chance nothing more needs to be done to get rid of the unnecessary `llvm` dependency.

![image](https://user-images.githubusercontent.com/24633722/207161778-9c7d46d5-0dd3-41fd-a9bb-6b0fcbc4bf97.png)

Reply:
Yes, after further investigation it seems that llvm is brought by mesa-dri-drivers with qt5 gui subpackage, Updated PR to simplify changes.